mlir/lib/Dialect/SCF/Transforms/ForallToFor.cpp

```diff
@@ -0,0 +1,90 @@
+//===- ForallToFor.cpp - scf.forall to scf.for loop conversion ------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Transforms SCF.ForallOp's into SCF.ForOp's.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/SCF/Transforms/Passes.h"
+
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+#define GEN_PASS_DEF_SCFFORALLTOFORLOOP
+#include "mlir/Dialect/SCF/Transforms/Passes.h.inc"
+} // namespace mlir
+
+using namespace llvm;
+using namespace mlir;
+using scf::ForallOp;
+using scf::ForOp;
+using scf::LoopNest;
+
+LogicalResult
+mlir::scf::forallToForLoop(RewriterBase &rewriter, scf::ForallOp forallOp,
+                           SmallVector<Operation *> *results = nullptr) {
+  rewriter.setInsertionPoint(forallOp);
+
+  if (!forallOp.getOutputs().empty()) {
+    return forallOp.emitOpError()
+           << "unsupported shared outputs (didn't bufferize?)";
+  }
+
+  auto loc = forallOp.getLoc();
+  SmallVector<Value> lbs = getValueOrCreateConstantIndexOp(
+      rewriter, loc, forallOp.getMixedLowerBound());
+  SmallVector<Value> ubs = getValueOrCreateConstantIndexOp(
+      rewriter, loc, forallOp.getMixedUpperBound());
+  SmallVector<Value> steps =
+      getValueOrCreateConstantIndexOp(rewriter, loc, forallOp.getMixedStep());
+  LoopNest loopNest = scf::buildLoopNest(rewriter, loc, lbs, ubs, steps);
+
+  if (results) {
+    llvm::copy(loopNest.loops, std::back_inserter(*results));
+  }
+
+  SmallVector<Value> ivs;
+  for (scf::ForOp loop : loopNest.loops) {
+    ivs.push_back(loop.getInductionVar());
+  }
+
+  Block *innermostBlock = loopNest.loops.back().getBody();
+  rewriter.eraseOp(forallOp.getBody()->getTerminator());
+  rewriter.inlineBlockBefore(forallOp.getBody(), innermostBlock,
+                             innermostBlock->getTerminator()->getIterator(),
+                             ivs);
+  rewriter.eraseOp(forallOp);
+
+  return success();
+}
+
+namespace {
+struct ForallToForLoop : public impl::SCFForallToForLoopBase<ForallToForLoop> {
+  void runOnOperation() override {
+    auto *parentOp = getOperation();
+    IRRewriter rewriter(parentOp->getContext());
+
+    SmallVector<scf::ForallOp> forallOps;
+    parentOp->walk(
+        [&](scf::ForallOp forallOp) { forallOps.push_back(forallOp); });
+
+    for (auto forallOp : forallOps) {
+      if (failed(scf::forallToForLoop(rewriter, forallOp))) {
+        return signalPassFailure();
+      }
+    }
```

Comment:
Is it not possible to call the rewrite function directly in the walk?

Reply:
I got this from the following comment: https://github.com/llvm/llvm-project/pull/89636#discussion_r1575670278

> Tip: during the walk, first do a walk to collect the ops in a SmallVector and then transform the ops.

Reply:
This "tip" unfortunately doesn't explain why ops need to be stored in a vector, which may not be necessary. Walk is a post-order traversal with early-increment iterator, so it should be perfectly fine to erase the current operation (increment already happened) and to move its body (that was already visited because post-order).

Reply:
I understood this as being a preferred way of doing this kind of rewrite. I don't have a preference either way, so I updated it to rewrite directly inside the walk as suggested here.

Reply:
The preferred way is specifically using the `walk` instead of `applyPatternsGreedily` for a small number of patterns as the latter comes with significant overhead and additional caveats (it will be doing constant folding and CFG simplification, for example).